### ISSUE TITLE
onAfter to support things like dumping chrome perf report

### ISSUE THREAD

**Issue by skarfacegc** [garris/BackstopJS#1077]
Since there was another ticket closed for an already supported use case I figured I'd open a request for onAfter for another use case that I'm not sure would be supported by putting cleanup at the start on the onReady/onBefore script.

I would love to use the [puppeteer trace](https://github.com/GoogleChrome/puppeteer/blob/v1.19.0/docs/api.md#class-tracing) to capture page load time reports along with the screenshot. AFAICT there's no way to do this with the existing Backstop API

**Comment by skarfacegc** [comment]
I've done a bit of playing with the backstop source to see if I could wedge this in.
I ran into puppeteer-har that seems to automate the setup of puppeteer trace.

I could find the place to issue the har.start  (had to attach the har object to the scenario object) but I couldn't figure out the right place to issue the hart.stop()

See example here https://github.com/Everettss/puppeteer-har

IMO it would be super useful to have the option to capture perf stats as well as images.

**Comment by garris** [comment]
Yes.  I was working with an engineer yesterday who had a similar challenge -- it would have been super useful to profile the network activity etc.  so, I agree that could be useful.  What would that look like?  How would users interact with backstop to use it?

**Comment by skarfacegc** [comment]
Turns out that page.tracing is different than HAR  (learning this as I go).  Seems like it would make sense to add a captureTrace and captureHAR options (or similar) to the scenario and save the appropriate trace files using the same naming scheme as the snapshots. I'm not sure that a simple onAfter hook would suffice.  (I could be wrong though, I've only played with the Backstop code a bit)

**Comment by garris** [comment]
For puppeteer, what do you need to call -- and when do you need to call it?

**Comment by skarfacegc** [comment]
Before the page.goto()
page.tracing.start({path: filename});

right before/after the screenshot
page.tracing.stop();

**Comment by garris** [comment]
If `page.tracing.stop();` can be executed prior to screenshot then this can simply be added to onReady.  Does this not work?

**Comment by skarfacegc** [comment]
Hrm, quite possibly.  I was going with the approach of going after the screenshot, but I don't know that it matters.  I'll give that a shot this evening and see if it works.

**Comment by skarfacegc** [comment]
Looks like adding page.tracing.start(....) to onBefore and page.tracing.stop() to onReady seems to work.  Do I have access to the file naming code inside of the onBefore so I can lineup the trace filenames with the image file names?

**Comment by garris** [comment]
Nice 👍 
Yes,  the name of the test should be attached to the scenario object.

**Comment by garris** [comment]
There may also be a property with a name that is safe for filenames.

**Comment by skarfacegc** [comment]
One last question ... I dug through the code a bit and figured out that the end of onBefore is really close to the page.goto.  What I couldn't figure out is how close is onReady to the capture?
